From 87992cb405a96e5762038ddad050bd8274042aa3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marc=20B=C3=A4chinger?= <bachinger@google.com>
Date: Wed, 10 Sep 2025 14:47:32 +0000
Subject: [PATCH 18/56] Check length of MBR component name properties

Test: Can't be tested on main as PackageManager prevents it
Flag: EXEMPT BUGFIX
Bug: 433250316
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:9062dd861d49627405804fcc9bb50fdae3701aa2
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:b769129921d76b8bd9ec7d2ed35e5044f362857d
Merged-In: Ib3f0ec4277a468e4eb9b76604d5affd2568e8a3d
Change-Id: Ib3f0ec4277a468e4eb9b76604d5affd2568e8a3d
---
 .../android/server/media/MediaButtonReceiverHolder.java  | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/media/MediaButtonReceiverHolder.java b/services/core/java/com/android/server/media/MediaButtonReceiverHolder.java
index 1f42d18b8..866651d4c 100644
--- a/services/core/java/com/android/server/media/MediaButtonReceiverHolder.java
+++ b/services/core/java/com/android/server/media/MediaButtonReceiverHolder.java
@@ -67,7 +67,7 @@ final class MediaButtonReceiverHolder {
     // Filter apps regardless of the phone's locked/unlocked state.
     private static final int PACKAGE_MANAGER_COMMON_FLAGS =
             PackageManager.MATCH_DIRECT_BOOT_AWARE | PackageManager.MATCH_DIRECT_BOOT_UNAWARE;
-
+    private static final int MAX_COMPONENT_NAME_LENGTH = 1024;
     private final int mUserId;
     private final PendingIntent mPendingIntent;
     private final ComponentName mComponentName;
@@ -403,6 +403,13 @@ final class MediaButtonReceiverHolder {
             if (componentInfo != null && TextUtils.equals(componentInfo.packageName,
                     pendingIntent.getCreatorPackage())
                     && componentInfo.packageName != null && componentInfo.name != null) {
+                int componentNameLength =
+                        componentInfo.packageName.length() + componentInfo.name.length() + 1;
+                if (componentNameLength > MAX_COMPONENT_NAME_LENGTH) {
+                    Log.w(TAG, "detected and ignored component name with overly long package"
+                            + " or name, pi=" + pendingIntent);
+                    continue;
+                }
                 return new ComponentName(componentInfo.packageName, componentInfo.name);
             }
         }
-- 
2.52.0

